board/drivers/fdcan.h

```diff
@@ -0,0 +1,203 @@
+
+    if ((CANx->TXFQS & FDCAN_TXFQS_TFQF) == 0) {
+      CAN_FIFOMailBox_TypeDef to_send;
+      if (can_pop(can_queues[bus_number], &to_send)) {
+        can_tx_cnt += 1;
+        uint32_t TxFIFOSA = FDCAN_START_ADDRESS + (can_number * FDCAN_OFFSET) + (FDCAN_RX_FIFO_0_EL_CNT * FDCAN_RX_FIFO_0_EL_SIZE);
+        uint8_t tx_index = (CANx->TXFQS >> FDCAN_TXFQS_TFQPI_Pos) & 0x1F;
+        // only send if we have received a packet
+        CAN_FIFOMailBox_TypeDef *fifo;
+        fifo = (CAN_FIFOMailBox_TypeDef *)(TxFIFOSA + (tx_index * FDCAN_TX_FIFO_EL_SIZE));
+
+        // Convert from "mailbox type"
+        fifo->RIR = ((to_send.RIR & 0x6) << 28) | (to_send.RIR >> 3);  // identifier format and frame type | identifier
+        //REDEBUG: enable CAN FD and BRS for test purposes
+        //fifo->RDTR = ((to_send.RDTR & 0xF) << 16) | ((to_send.RDTR) >> 16) | (1U << 21) | (1U << 20); // DLC (length) | timestamp | enable CAN FD | enable BRS
+        fifo->RDTR = ((to_send.RDTR & 0xF) << 16) | ((to_send.RDTR) >> 16); // DLC (length) | timestamp
+        fifo->RDLR = to_send.RDLR;
+        fifo->RDHR = to_send.RDHR;
+        
+        CANx->TXBAR = (1UL << tx_index); 
+
+        // Send back to USB
+        can_txd_cnt += 1;
+        CAN_FIFOMailBox_TypeDef to_push;
+        to_push.RIR = to_send.RIR;
+        to_push.RDTR = (to_send.RDTR & 0xFFFF000FU) | ((CAN_BUS_RET_FLAG | bus_number) << 4);
+        to_push.RDLR = to_send.RDLR;
+        to_push.RDHR = to_send.RDHR;
+        can_send_errs += can_push(&can_rx_q, &to_push) ? 0U : 1U;
+
+        if (can_tx_check_min_slots_free(MAX_CAN_MSGS_PER_BULK_TRANSFER)) {
+          usb_outep3_resume_if_paused();
+        }
+      }
+    }
+
+    // Recover after Bus-off state
+    if (((CANx->PSR & FDCAN_PSR_BO) != 0) && ((CANx->CCCR & FDCAN_CCCR_INIT) != 0)) {
+      bus_off_err[can_number] += 1U;
+      puts("CAN is in Bus_Off state! Resetting... CAN number: "); puth(can_number); puts("\n");
+      if (bus_off_err[can_number] > BUS_OFF_FAIL_LIMIT) {
+        cycle_transceiver(can_number);
+      }
+      CANx->IR = 0xFFC60000U; // Reset all flags(Only errors!)
+      CANx->CCCR &= ~(FDCAN_CCCR_INIT);
+      uint32_t timeout_counter = 0U;
+      while((CANx->CCCR & FDCAN_CCCR_INIT) != 0) {
+        // Delay for about 1ms
+        delay(10000);
+        timeout_counter++;
+
+        if(timeout_counter >= CAN_INIT_TIMEOUT_MS){
+          puts(CAN_NAME_FROM_CANIF(CANx)); puts(" Bus_Off reset timed out!\n");
+          break;
+        }
+      }
+    }
+    EXIT_CRITICAL();
+  }
+}
+
+// CAN receive handlers
+// blink blue when we are receiving CAN messages
+void can_rx(uint8_t can_number) {
+  FDCAN_GlobalTypeDef *CANx = CANIF_FROM_CAN_NUM(can_number);
+  uint8_t bus_number = BUS_NUM_FROM_CAN_NUM(can_number);
+	uint8_t rx_fifo_idx;
+
+	// Rx FIFO 0 new message
+	if((CANx->IR & FDCAN_IR_RF0N) != 0) {
+    CANx->IR |= FDCAN_IR_RF0N;
+    while((CANx->RXF0S & FDCAN_RXF0S_F0FL) != 0) {
+      can_rx_cnt += 1;
```

Comment:
wrong indentation level

Reply:
Checked in 3 different editors - can't see the problem, but definitely see it on the github...